bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/EntryLogTest.java

```diff
@@ -391,4 +400,137 @@ public void testGetEntryLogsSet() throws Exception {
 
         assertEquals(Sets.newHashSet(0L, 1L, 2L, 3L), logger.getEntryLogsSet());
     }
+
+    static class LedgerStorageWriteTask implements Callable<Boolean> {
+        long ledgerId;
+        int entryId;
+        LedgerStorage ledgerStorage;
+
+        LedgerStorageWriteTask(long ledgerId, int entryId, LedgerStorage ledgerStorage) {
+            this.ledgerId = ledgerId;
+            this.entryId = entryId;
+            this.ledgerStorage = ledgerStorage;
+        }
+
+        @Override
+        public Boolean call() {
+            try {
+                ledgerStorage.addEntry(generateEntry(ledgerId, entryId));
+            } catch (IOException e) {
+                LOG.error("Got Exception for AddEntry call. LedgerId: " + ledgerId + " entryId: " + entryId, e);
+                return false;
+            }
+            return true;
+        }
+    }
+
+    static class LedgerStorageReadTask implements Callable<Boolean> {
+        long ledgerId;
+        int entryId;
+        LedgerStorage ledgerStorage;
+
+        LedgerStorageReadTask(long ledgerId, int entryId, LedgerStorage ledgerStorage) {
+            this.ledgerId = ledgerId;
+            this.entryId = entryId;
+            this.ledgerStorage = ledgerStorage;
+        }
+
+        @Override
+        public Boolean call() {
+            try {
+                String expectedValue = generateDataString(ledgerId, entryId);
+                ByteBuf byteBuf = ledgerStorage.getEntry(ledgerId, entryId);
+                long actualLedgerId = byteBuf.readLong();
+                long actualEntryId = byteBuf.readLong();
+                byte[] data = new byte[byteBuf.readableBytes()];
+                byteBuf.readBytes(data);
+                if (ledgerId != actualLedgerId) {
+                    LOG.error("For ledgerId: {} entryId: {} readRequest, actual ledgerId: {}", ledgerId, entryId,
+                            actualLedgerId);
+                    return false;
+                }
+                if (entryId != actualEntryId) {
+                    LOG.error("For ledgerId: {} entryId: {} readRequest, actual entryId: {}", ledgerId, entryId,
+                            actualEntryId);
+                    return false;
+                }
+                if (!expectedValue.equals(new String(data))) {
+                    LOG.error("For ledgerId: {} entryId: {} readRequest, actual Data: {}", ledgerId, entryId,
+                            new String(data));
+                    return false;
+                }
+            } catch (IOException e) {
+                LOG.error("Got Exception for GetEntry call. LedgerId: " + ledgerId + " entryId: " + entryId, e);
+                return false;
+            }
+            return true;
+        }
+    }
+
+    /**
+     * test concurrent write operations and then concurrent read
+     * operations using InterleavedLedgerStorage.
+     */
+    @Test
+    public void testConcurrentWriteAndReadCallsOfInterleavedLedgerStorage() throws Exception {
```

Comment:
This change effects SortedLedgerStorage too. Add a variant of the test that tests against that.

Reply:
why? I don't understand why you need a sorted ledger storage test here. the change is in entry log itself. the entry log is used by all ledger storage implementation, that doesn't mean you need to test all ledger storage implementation, which is redundant. the test should focus on testing what it is changing, not testing all the variants that is calling this class, otherwise we will end up have lots of redundant tests testing same logic, which make the whole build time very long.

so I don't think we need redundant tests here.

Reply:
> I don't understand why you need a sorted ledger storage test here. the change is in entry log itself.

This change is in interleavedledgerstorage, which sortedledgerstorage derives from. It removes synchronization from processEntry, which sortedledgerstorage calls directly. It seems there's no other synchronization in sortedledgerstorage though, so it should be ok.

> the test should focus on testing what it is changing, 

which is _exactly_ why we shouldn't be constructing a full bookie to test it.

Reply:
> This change is in interleavedledgerstorage, which sortedledgerstorage derives from. It removes synchronization from processEntry, which sortedledgerstorage calls directly. It seems there's no other synchronization in sortedledgerstorage though, so it should be ok.

this synchronization change only makes sense at InterleavedLedgerStorage, because there can be concurrent calls to `processEntry` when using InterleavedLedgerStorage. However this is not true in SortedLedgerStorage, because ledger storage flushes entries in a single-threaded executor. so I don't see why you need a test from SortedLedgerStorage. 

The test case is testing concurrent reads and write on EntryLogTest, InterleavedLedgerStorage is the tool to use for testing. It is not testing concurrent reads and writes on ledger storage. 

so 1) the testing object is EntryLog not ledger storage 2) concurrent reads and writes only make sense in interleaved ledger storage. so it is not right to add a duplicated test using sorted ledger storage.

> which is exactly why we shouldn't be constructing a full bookie to test it.

that's different. what I am talking here is you need to know what you are exactly _testing_. as I commented above, the testing subject is EntryLog not LedgerStorage.

whether constructing a full bookie or constructing a ledger storage is just a tool to setup a test case for testing EntryLog. 

the whole test suite is using Bookie already. for consistency, it is okay to use Bookie here. I don't see a strong reason to block this change just because of that.

Reply:
> whether constructing a full bookie or constructing a ledger storage is just a tool to setup a test case for testing EntryLog.

If it's only testing EntryLogger, then only entrylogger should be constructed.

But the change has changed flushing also, and removed locking that was previous around access to the index, so the how ledger storage should be tested. 

> the whole test suite is using Bookie already. for consistency, it is okay to use Bookie here. I don't see a strong reason to block this change just because of that.

The rest of the suite is testing badly. We shouldn't propagate bad practices. I don't consider consistency a good argument against this.

Reply:
@reddycharan it would be good to have a thread running in a tight loop calling flush() on the ledgerstorage, while the adds and reads are happening.

Reply:
> If it's only testing EntryLogger, then only entrylogger should be constructed.

in order to make test entry logger, you need to setup all the facility for do concurrent writes and reads. and InterleavedLedgerStorage is the facility. what is wrong with this? It is a granularity problem on the testing facility, if you just just using EntryLogger you need to setup too many additional stuffs for testing.

> But the change has changed flushing also, and removed locking that was previous around access to the index, so the how ledger storage should be tested.

yes. that's why it is using InterleaveLedgerStorage. because it is a matter for InterleaveLedgerStorage. it is not a matter of SortedLedgerStorage.

Reply:
> @reddycharan it would be good to have a thread running in a tight loop calling flush() on the ledgerstorage, while the adds and reads are happening.

fair enough I can add flush tasks  to the write and read tasks list, so that flush would be executed along with write/read operations.